# **OBJECTIVE FUNCTIONS:**

#### Objective functions, also known as loss functions or cost functions, are crucial in machine learning and statistics because they provide a measure of how well a model's predictions align with the actual data.

### Some of the widely used fuctions are:
#### - RMSE [ Root Mean Squared Error]
#### - MAE [ Mean Absolute Error ]
#### - Logistic Loss
#### - Cross Entropy

### **Will Discuss these one by one**



### 1. RMSE (Root Mean Squared Error)

**Use case:** Regression

**Formula:**
$$
\text{RMSE} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

**Explanation:**

- **$y_i$**: The true value for the $i$-th observation.
- **$\hat{y}_i$**: The predicted value for the $i-th$ observation.
- **$n$**: The number of observations.

**Details:**

- RMSE measures the square root of the average squared differences between the predicted values and the actual values.
- It gives higher weight to larger errors, meaning it is sensitive to outliers.
- A lower RMSE value indicates a better fit of the model to the data.

**Interpretation:**

- RMSE is in the same units as the response variable $y$.
- It is useful for comparing the predictive accuracy of different models on the same dataset.


In [ ]:
import pandas as pd
import numpy as np
NY_data = pd.read_csv('NYCTaxiFareData.csv')

In [ ]:
NY_data.head()

key  fare_amount          pickup_datetime  \
0    2009-06-15 17:26:21.0000001          4.5  2009-06-15 17:26:21 UTC   
1    2010-01-05 16:52:16.0000002         16.9  2010-01-05 16:52:16 UTC   
2   2011-08-18 00:35:00.00000049          5.7  2011-08-18 00:35:00 UTC   
3    2012-04-21 04:30:42.0000001          7.7  2012-04-21 04:30:42 UTC   
4  2010-03-09 07:51:00.000000135          5.3  2010-03-09 07:51:00 UTC   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.844311        40.721319         -73.841610         40.712278   
1        -74.016048        40.711303         -73.979268         40.782004   
2        -73.982738        40.761270         -73.991242         40.750562   
3        -73.987130        40.733143         -73.991567         40.758092   
4        -73.968095        40.768008         -73.956655         40.783762   

   passenger_count  
0                1  
1                1  
2                2  
3                1  
4                1

In [ ]:
NY_data.describe()

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  10000.000000      10000.000000     10000.000000       10000.000000   
mean      11.235464        -72.466660        39.920448         -72.474094   
std        9.584258         10.609729         7.318932          10.579732   
min       -2.900000        -74.438233       -74.006893         -74.429332   
25%        6.000000        -73.992058        40.734547         -73.991112   
50%        8.500000        -73.981758        40.752693         -73.980083   
75%       12.500000        -73.966925        40.767694         -73.963504   
max      180.000000         40.766125       401.083332          40.802437   

       dropoff_latitude  passenger_count  
count      10000.000000     10000.000000  
mean          39.893281         1.644700  
std            6.339919         1.271229  
min          -73.994392         0.000000  
25%           40.735230         1.000000  
50%           40.753738         1.000000  
75%           40.768186         2.000000  
max           41.366138         6.000000

In [ ]:
NY_data['pickup_datetime'] = pd.to_datetime(NY_data['pickup_datetime'])

# Extract useful datetime features
NY_data['year'] = NY_data['pickup_datetime'].dt.year
NY_data['month'] = NY_data['pickup_datetime'].dt.month
NY_data['day'] = NY_data['pickup_datetime'].dt.day
NY_data['hour'] = NY_data['pickup_datetime'].dt.hour

NY_data = NY_data.drop(columns=['pickup_datetime'])

In [ ]:
NY_data

key  fare_amount  pickup_longitude  \
0       2009-06-15 17:26:21.0000001          4.5        -73.844311   
1       2010-01-05 16:52:16.0000002         16.9        -74.016048   
2      2011-08-18 00:35:00.00000049          5.7        -73.982738   
3       2012-04-21 04:30:42.0000001          7.7        -73.987130   
4     2010-03-09 07:51:00.000000135          5.3        -73.968095   
...                             ...          ...               ...   
9995   2011-10-26 10:44:00.00000086         11.7        -73.988277   
9996  2011-12-16 15:37:00.000000179          5.7        -74.002112   
9997    2013-11-16 22:47:17.0000001         12.0        -73.992093   
9998   2010-01-28 11:38:00.00000022          6.5        -73.992548   
9999    2011-07-05 06:48:44.0000002         10.5        -73.978400   

      pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0           40.721319         -73.841610         40.712278                1   
1           40.711303         -73.979268         40.782004                1   
2           40.761270         -73.991242         40.750562                2   
3           40.733143         -73.991567         40.758092                1   
4           40.768008         -73.956655         40.783762                1   
...               ...                ...               ...              ...   
9995        40.748970         -73.963712         40.773958                2   
9996        40.748727         -73.992467         40.756252                1   
9997        40.729071         -73.974470         40.763050                2   
9998        40.735652         -73.998802         40.723085                1   
9999        40.751700         -74.010500         40.713600                1   

      year  month  day  hour  
0     2009      6   15    17  
1     2010      1    5    16  
2     2011      8   18     0  
3     2012      4   21     4  
4     2010      3    9     7  
...    ...    ...  ...   ...  
9995  2011     10   26    10  
9996  2011     12   16    15  
9997  2013     11   16    22  
9998  2010      1   28    11  
9999  2011      7    5     6  

[10000 rows x 11 columns]

In [ ]:
X = NY_data.drop(columns=['fare_amount', 'key'])
y = NY_data['fare_amount']

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f'RMSE: {rmse}')

RMSE: 10.055300400668683


### Understanding RMSE

* An RMSE of 10.055 means that, on average, the model's predictions are about 10.055 units away from the actual values.

* The significance of an RMSE value depends on the context and scale of the data. For example, in a fare prediction model, if the fare amounts range from 0 to 100, an RMSE of 10.055 might be considered high. However, if fare amounts range from 0 to 1000, an RMSE of 10.055 might be considered low.

To improve RMSE, consider more advanced models, feature engineering, or better data preprocessing techniques. Will Discuss in further sections


### 2. MAE (Mean Absolute Error)

**Use case:** Regression

**Formula:**

$$\text{MAE} = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|$$

**Explanation:**

- **$y_i$**: The true value for the $i-th$ observation.
- **$\hat{y}_i$**: The predicted value for the $i-th$ observation.
- **$n$**: The number of observations.

**Details:**

- MAE measures the average magnitude of the errors in a set of predictions, without considering their direction.
- It is the average of the absolute differences between the predicted values and the actual values.
- MAE treats all errors equally.

**Interpretation:**

- MAE is in the same units as the response variable $y$.
- It is useful for understanding the typical size of the errors in predictions.



In [ ]:
mae = mean_absolute_error(y_test, y_pred)
print(f'MAE: {mae}')

MAE: 6.054968368599895



### 3. Logistic Loss (Binary Cross-Entropy Loss)

**Use case:** Binary classification

**Formula:**

$$text{Logistic Loss} = -\frac{1}{n} \sum_{i=1}^{n} [y_i \log(\hat{y}_i) + (1 - y_i) \log(1 - \hat{y}_i)] $$

**Explanation:**

- **$y_i$**: The true binary label (0 or 1) for the $i-th$ observation.
- **$\hat{y}_i$**: The predicted probability that the $i-th$ observation belongs to class 1.
- **$n$**: The number of observations.

**Details:**

- Logistic loss measures the performance of a classification model where the output is a probability value between 0 and 1.
- It penalizes false classifications. A higher probability assigned to the correct class results in a lower loss.
- The loss increases as the predicted probability diverges from the actual label.

**Interpretation:**

- A lower logistic loss indicates better performance of the classifier.
- It is particularly useful for models that output probabilities, such as logistic regression.


In [ ]:
TitanicData= pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

In [ ]:
TitanicData.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
TitanicData = TitanicData.drop(columns=['Name', 'Ticket', 'Cabin'])

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [ ]:
TitanicData['Age'].fillna(TitanicData['Age'].median(), inplace=True)
TitanicData['Embarked'].fillna(TitanicData['Embarked'].mode()[0], inplace=True)

/var/folders/9s/jhct1nvn5w551vq58kr4g2rm0000gn/T/ipykernel_9788/215853253.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  TitanicData['Age'].fillna(TitanicData['Age'].median(), inplace=True)
/var/folders/9s/jhct1nvn5w551vq58kr4g2rm0000gn/T/ipykernel_9788/215853253.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are set

In [ ]:
X = TitanicData.drop(columns=['Survived'])
y = TitanicData['Survived']

In [ ]:
numeric_features = ['Age', 'Fare']
numeric_transformer = StandardScaler()

categorical_features = ['Sex', 'Embarked', 'Pclass']
categorical_transformer = OneHotEncoder(drop='first')

In [ ]:
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

In [ ]:
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression())
])

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', StandardScaler(),
                                                  ['Age', 'Fare']),
                                                 ('cat',
                                                  OneHotEncoder(drop='first'),
                                                  ['Sex', 'Embarked',
                                                   'Pclass'])])),
                ('classifier', LogisticRegression())])

In [ ]:
y_pred_proba = pipeline.predict_proba(X_test)
loss = log_loss(y_test, y_pred_proba)

print(f'Logistic Loss: {loss}')

Logistic Loss: 0.43295800972366816


### 4. Cross Entropy Loss (Categorical Cross-Entropy Loss)

**Use case:** Multi-class classification

**Formula:**

$$\text{Cross Entropy} = -\sum_{i=1}^{n} \sum_{c=1}^{k} y_{i,c} \log(\hat{y}_{i,c})$$

**Explanation:**

- **$n$**: The number of observations.
- **$k$**: The number of classes.
- **$y_{i,c}$**: A binary indicator (0 or 1) if class label $c$ is the correct classification for observation $i$.
- **$\hat{y}_{i,c}$**: The predicted probability that observation $i$ belongs to class $c$.

**Details:**

- Cross entropy loss measures the performance of a classification model where the output is a probability distribution over multiple classes.
- It is the sum of the negative log probabilities of the true class labels.
- It penalizes the probability of the correct class being low.

**Interpretation:**

- A lower cross entropy loss indicates a better performance of the classifier.
- It is especially useful for multi-class classification problems where the output is a probability distribution.

### Summary of Interpretations:

- **RMSE and MAE** are used for regression tasks, with RMSE being more sensitive to outliers.
- **Logistic loss** is used for binary classification and measures the probability predictions against the true binary labels.
- **Cross entropy loss** is used for multi-class classification and measures the probability distributions against the true class labels.

Each of these objective functions helps quantify how well a model's predictions match the actual data, guiding the optimization process during model training.

In [ ]:
pip install tensorflow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.4/45.4 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 227.0/227.0 MB 11.0 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.7/133.7 kB 18.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 26.8 MB/s eta 0:00:0000:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 44.1 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 49.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.4/26.4 MB 37.3 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 389.8/389.8 kB 25.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 52.2 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 52.3 MB/s eta 0:00:00a 0:00:01

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.losses import CategoricalCrossentropy

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 1s 0us/step


In [3]:
X_train = X_train / 255.0
X_test = X_test / 255.0

In [4]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [5]:
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

In [6]:
model.compile(optimizer='adam',
              loss=CategoricalCrossentropy(),
              metrics=['accuracy'])

In [7]:
model.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.2)

Epoch 1/5
1500/1500 [==============================] - 13s 7ms/step - loss: 0.2671 - accuracy: 0.9222 - val_loss: 0.1351 - val_accuracy: 0.9601
Epoch 2/5
1500/1500 [==============================] - 7s 5ms/step - loss: 0.1153 - accuracy: 0.9655 - val_loss: 0.1084 - val_accuracy: 0.9679
Epoch 3/5
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0784 - accuracy: 0.9764 - val_loss: 0.0943 - val_accuracy: 0.9710
Epoch 4/5
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0577 - accuracy: 0.9820 - val_loss: 0.0963 - val_accuracy: 0.9697
Epoch 5/5
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0457 - accuracy: 0.9854 - val_loss: 0.0869 - val_accuracy: 0.9741


In [8]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Categorical Cross-Entropy Loss: {loss}')
print(f'Accuracy: {accuracy}')

313/313 [==============================] - 1s 3ms/step - loss: 0.0780 - accuracy: 0.9759
Categorical Cross-Entropy Loss: 0.0780465304851532
Accuracy: 0.9758999943733215


------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

With this, we will know if out model is doing or do we have to do some feature engineering steps; or if we have to use other models; So these **Objective Functions** are very important

There are other functions than these like:

### For Regression:
1. **Mean Squared Error (MSE)**:
  $$
   \text{MSE} = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2
   $$
   - Similar to RMSE but without taking the square root. It also emphasizes larger errors.

2. **Huber Loss**:
   $$
   L_\delta(a) = \begin{cases} 
   \frac{1}{2}a^2 & \text{for } |a| \leq \delta, \\
   \delta(|a| - \frac{1}{2}\delta) & \text{for } |a| > \delta,
   \end{cases}
   $$
   where $a = y_i - \hat{y}_i$.
   - Combines the best properties of MAE and MSE, being less sensitive to outliers than MSE but more sensitive than MAE.

3. **Quantile Loss**:
   $$
   \text{Quantile Loss} = \sum_{i=1}^{n} \left( \tau (y_i - \hat{y}_i)^+ + (1-\tau) (\hat{y}_i - y_i)^+ \right)
   $$
   - Used in quantile regression to predict the quantiles (e.g., median) of the target variable.

### For Classification:
1. **Hinge Loss (for Support Vector Machines)**:
   $$
   \text{Hinge Loss} = \frac{1}{n} \sum_{i=1}^{n} \max(0, 1 - y_i \cdot \hat{y}_i)
   $$
   - Used for training SVMs, focusing on correctly classifying samples with a margin.

2. **Kullback-Leibler Divergence (KL Divergence)**:
   $$
   \text{KL}(P || Q) = \sum_{i} P(i) \log \left( \frac{P(i)}{Q(i)} \right)
   $$
   - Measures how one probability distribution diverges from a second, expected probability distribution.

3. **Focal Loss**:
   $$
   \text{Focal Loss} = -\sum_{i=1}^{n} (1 - \hat{p}_{i})^\gamma \log(\hat{p}_{i})
   $$
   - Used to address class imbalance by focusing on hard-to-classify examples.

4. **Cosine Similarity Loss**:
   $$
   \text{Cosine Similarity Loss} = 1 - \cos(\theta) = 1 - \frac{A \cdot B}{\|A\| \|B\|}
   $$
   - Measures the cosine of the angle between two vectors, used in tasks like text similarity.

### For Sequence Models and Others:
1. **Connectionist Temporal Classification (CTC) Loss**:
   $$
   \text{CTC Loss} = -\log P(y|x)
   $$
   - Used in sequence modeling tasks where the alignment between input and output sequences is unknown (e.g., speech recognition).

2. **Earth Mover's Distance (Wasserstein Loss)**:
   $$
   W(p, q) = \inf_{\gamma \in \Pi(p, q)} \mathbb{E}_{(x,y) \sim \gamma} [ \|x - y\| ]
   $$
   - Measures the minimum cost of transforming one distribution into another, useful in generative models.

### For Ranking Problems:
1. **Pairwise Ranking Loss**:
  $$
   L = \sum_{i, j} \max(0, 1 - (s_i - s_j))
   $$
   - Used in ranking tasks where the goal is to order items correctly.

2. **Listwise Loss (e.g., ListNet)**:
   $$
   L = - \sum_{i} P(y_i) \log(Q(y_i))
  $$
   - Used in learning to rank tasks, considering the entire list of items.

### Regularization Terms:
While not objective functions by themselves, regularization terms are often added to objective functions to prevent overfitting:

1. **L1 Regularization (Lasso)**:
   $$
   \text{L1} = \lambda \sum_{j=1}^{p} |w_j|
   $$
   - Encourages sparsity in the model parameters.

2. **L2 Regularization (Ridge)**:
   $$
   \text{L2} = \lambda \sum_{j=1}^{p} w_j^2
   $$
   - Encourages smaller model parameters, leading to simpler models.

### Conclusion:
The choice of objective function depends on the specific characteristics of the problem being solved, the type of data, and the desired properties of the solution. The objective functions mentioned above are just some of the many options available, each suited to different types of tasks and model requirements.